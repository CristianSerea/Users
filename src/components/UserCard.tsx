import React, {memo, useMemo} from 'react';
import {Text, View, StyleSheet, TouchableOpacity} from 'react-native';
import useColorsTheme from '../constants/colors';
import {dimensions} from '../constants/dimensions';
import {fonts} from '../constants/fonts';
import {User} from '../types/types';
import Strings from '../../strings/strings.json';

type Props = {
  user: User;
  index: number;
  onPress: () => void;
};

const getFullAddress = (user: User) => {
  return (
    user.address.street + ' ' + user.address.city + ' ' + user.address.zipcode
  );
};

const UserCard = ({user, index, onPress}: Props) => {
  const {colors, shadow} = useColorsTheme();

  const marginRight = useMemo(() => {
    return index % 2 === 1 ? dimensions.margin : 0;
  }, [index]);

  const getAttribute = (title: string, attribute: string) => {
    return (
      <View style={styles.atrribute} testID={`${title}-${attribute}`}>
        <Text style={[fonts.normal]}>{title}:</Text>
        <Text style={[fonts.body]} testID={`${attribute}`}>
          {attribute}
        </Text>
      </View>
    );
  };

  return (
    <TouchableOpacity
      style={[
        styles.card,
        shadow,
        {backgroundColor: colors.whiteColor, marginRight: marginRight},
      ]}
      onPress={onPress}
      testID={`UserCard-${user.id}`}>
      <Text style={fonts.title}>{user.name}</Text>
      {getAttribute(Strings.Email, user.email)}
      {getAttribute(Strings.Phone, user.phone)}
      {getAttribute(Strings.Address, getFullAddress(user))}
      {getAttribute(Strings.Company, user.company.name)}
    </TouchableOpacity>
  );
};

export default memo(UserCard);

const styles = StyleSheet.create({
  card: {
    flex: 1,
    borderRadius: dimensions.borderRadius,
    padding: dimensions.margin / 2,
    marginBottom: dimensions.margin,
    marginLeft: dimensions.margin,
  },
  atrribute: {
    marginTop: dimensions.margin / 2,
  },
});
