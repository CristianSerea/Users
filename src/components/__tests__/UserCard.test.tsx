import React from 'react';
import '@testing-library/jest-native/extend-expect';
import {render, fireEvent} from '@testing-library/react-native';
import UserCard from '../UserCard';
import {User} from '../../types/types';
import users from '../../../mock/users.json';

const indexMock: number = 0;
const userMock: User = users[indexMock];

describe('UserCard tests', () => {
  it('Test user card rendering', () => {
    const {getByTestId, toJSON} = render(
      <UserCard user={userMock} index={indexMock} onPress={() => {}} />,
    );

    expect(getByTestId(`UserCard-${userMock.id}`)).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('Test user card attributes', () => {
    const {getByText, getByTestId} = render(
      <UserCard user={userMock} index={indexMock} onPress={() => {}} />,
    );

    expect(getByText(userMock.name)).toBeTruthy();

    expect(getByTestId(`Email-${userMock.email}`)).toBeTruthy();
    expect(getByTestId(`${userMock.email}`)).toHaveTextContent(userMock.email);

    expect(getByTestId(`Phone-${userMock.phone}`)).toBeTruthy();
    expect(getByTestId(`${userMock.phone}`)).toHaveTextContent(userMock.phone);

    const address = `${userMock.address.street} ${userMock.address.city} ${userMock.address.zipcode}`;
    expect(getByTestId(`Address-${address}`)).toBeTruthy();
    expect(getByTestId(`${address}`)).toHaveTextContent(address);

    expect(getByTestId(`Company-${userMock.company.name}`)).toBeTruthy();
    expect(getByTestId(`${userMock.company.name}`)).toHaveTextContent(
      userMock.company.name,
    );
  });

  it('Test user card action', () => {
    const onPressMock = jest.fn();
    const {getByTestId} = render(
      <UserCard user={userMock} index={indexMock} onPress={onPressMock} />,
    );

    fireEvent.press(getByTestId(`UserCard-${userMock.id}`));
    expect(onPressMock).toHaveBeenCalledTimes(1);
  });
});
