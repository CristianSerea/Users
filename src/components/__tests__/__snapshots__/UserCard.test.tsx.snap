// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard tests Renders user card 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 10,
      "elevation": 5,
      "flex": 1,
      "marginBottom": 20,
      "marginLeft": 20,
      "marginRight": 0,
      "opacity": 1,
      "padding": 10,
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 2,
        "width": 2,
      },
      "shadowOpacity": 0.5,
      "shadowRadius": 5,
    }
  }
  testID="UserCard-1"
>
  <Text
    style={
      {
        "fontSize": 18,
        "fontWeight": "bold",
      }
    }
  >
    Leanne Graham
  </Text>
  <View
    style={
      {
        "marginTop": 10,
      }
    }
    testID="Email-Sincere@april.biz"
  >
    <Text
      style={
        [
          {
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    >
      Email
      :
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 15,
          },
        ]
      }
      testID="Sincere@april.biz"
    >
      Sincere@april.biz
    </Text>
  </View>
  <View
    style={
      {
        "marginTop": 10,
      }
    }
    testID="Phone-1-770-736-8031 x56442"
  >
    <Text
      style={
        [
          {
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    >
      Phone
      :
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 15,
          },
        ]
      }
      testID="1-770-736-8031 x56442"
    >
      1-770-736-8031 x56442
    </Text>
  </View>
  <View
    style={
      {
        "marginTop": 10,
      }
    }
    testID="Address-Kulas Light Gwenborough 92998-3874"
  >
    <Text
      style={
        [
          {
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    >
      Address
      :
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 15,
          },
        ]
      }
      testID="Kulas Light Gwenborough 92998-3874"
    >
      Kulas Light Gwenborough 92998-3874
    </Text>
  </View>
  <View
    style={
      {
        "marginTop": 10,
      }
    }
    testID="Company-Romaguera-Crona"
  >
    <Text
      style={
        [
          {
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    >
      Company
      :
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 15,
          },
        ]
      }
      testID="Romaguera-Crona"
    >
      Romaguera-Crona
    </Text>
  </View>
</View>
`;

exports[`UserCard tests Test user card rendering 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 10,
      "elevation": 5,
      "flex": 1,
      "marginBottom": 20,
      "marginLeft": 20,
      "marginRight": 0,
      "opacity": 1,
      "padding": 10,
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 2,
        "width": 2,
      },
      "shadowOpacity": 0.5,
      "shadowRadius": 5,
    }
  }
  testID="UserCard-1"
>
  <Text
    style={
      {
        "fontSize": 18,
        "fontWeight": "bold",
      }
    }
  >
    Leanne Graham
  </Text>
  <View
    style={
      {
        "marginTop": 10,
      }
    }
    testID="Email-Sincere@april.biz"
  >
    <Text
      style={
        [
          {
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    >
      Email
      :
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 15,
          },
        ]
      }
      testID="Sincere@april.biz"
    >
      Sincere@april.biz
    </Text>
  </View>
  <View
    style={
      {
        "marginTop": 10,
      }
    }
    testID="Phone-1-770-736-8031 x56442"
  >
    <Text
      style={
        [
          {
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    >
      Phone
      :
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 15,
          },
        ]
      }
      testID="1-770-736-8031 x56442"
    >
      1-770-736-8031 x56442
    </Text>
  </View>
  <View
    style={
      {
        "marginTop": 10,
      }
    }
    testID="Address-Kulas Light Gwenborough 92998-3874"
  >
    <Text
      style={
        [
          {
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    >
      Address
      :
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 15,
          },
        ]
      }
      testID="Kulas Light Gwenborough 92998-3874"
    >
      Kulas Light Gwenborough 92998-3874
    </Text>
  </View>
  <View
    style={
      {
        "marginTop": 10,
      }
    }
    testID="Company-Romaguera-Crona"
  >
    <Text
      style={
        [
          {
            "fontSize": 15,
            "fontWeight": "600",
          },
        ]
      }
    >
      Company
      :
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 15,
          },
        ]
      }
      testID="Romaguera-Crona"
    >
      Romaguera-Crona
    </Text>
  </View>
</View>
`;
